### PR TITLE
Concatenate DWI

### PR DESCRIPTION
**Pull request by frheault**
New utility script.

This is a 'rarer', I didn't code any lazy saving (meaning if you fuse 100 DWI the final results is all in memory)
I also didn't code an approx. of bvals (meaning if you fuse 2 DWI with bval at 1001 and 999 they will be kept as is)
I also didn't check the data_type BUT there is a parameters that forces it.

### CODE REVIEW

File: scripts/scil_concatenate_dwi.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    all_bvecs = np.concatenate(all_bvecs)
+
+    ref_dwi = nib.load(args.in_dwi[0])
+    all_dwi = np.zeros(ref_dwi.shape[0:3] + (total_size,),
+                       dtype=args.data_type)
+    last_count = ref_dwi.shape[-1]
+    all_dwi[..., 0:last_count] = ref_dwi.get_fdata()
+    for i in range(1, len(args.in_dwi)):
+        curr_dwi = nib.load(args.in_dwi[i])
+        if curr_dwi.shape != ref_dwi.shape:
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:80]: if the forth dimension is different this will crash. no ?

File: scripts/scil_concatenate_dwi.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            raise ValueError('All DWI must have the same shape.')
+        curr_size = curr_dwi.shape[-1]
+        all_dwi[..., last_count:last_count+curr_size] = curr_dwi.get_fdata()
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:83]: please add dtype.

File: scripts/scil_concatenate_dwi.py (hunk 1 above)
Comment (arnaudbore) [reply scripts/scil_concatenate_dwi.py:80]: We should also add is_header_compatible

File: scripts/scil_concatenate_dwi.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('out_dwi',
+                   help='The name of the output DWI file.')
+    p.add_argument('out_bvals',
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:27]: out_bval and out_bvec (just one file)
```suggestion
    p.add_argument('out_bval',
```

File: scripts/scil_concatenate_dwi.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                   help='The name of the output b-values.')
+    p.add_argument('out_bvecs',
+                   help='The name of the output b-vectors')
+
+    p.add_argument('--in_dwi', nargs='+',
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:32]:
```suggestion
    p.add_argument('--in_dwis', nargs='+',
```

File: scripts/scil_concatenate_dwi.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    p = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('out_dwi',
+                   help='The name of the output DWI file.')
+    p.add_argument('out_bvals',
+                   help='The name of the output b-values.')
+    p.add_argument('out_bvecs',
+                   help='The name of the output b-vectors')
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:30]:
```suggestion
                   help='The name of the output b-vectors.')
```

File: scripts/scil_concatenate_dwi.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+def main():
+    parser = _build_arg_parser()
+    args = parser.parse_args()
+
+    if len(args.in_dwi) != len(args.in_bvals) \
+            or len(args.in_dwi) != len(args.in_bvecs):
+        parser.error('DWI, bvals and bvecs must have the same length')
+
+    assert_inputs_exist(parser, args.in_dwi + args.in_bvals + args.in_bvecs)
+    assert_outputs_exist(parser, args, [args.out_dwi, args.out_bvals,
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:57]:
```suggestion
    assert_outputs_exist(parser, args, [args.out_dwi, args.out_bval,
```

File: scripts/scil_concatenate_dwi.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                                        args.out_bvecs])
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:58]:
```suggestion
                                        args.out_bvec])
```

File: scripts/scil_concatenate_dwi.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    all_bvals = []
+    all_bvecs = []
+    total_size = 0
+    for i in range(len(args.in_dwi)):
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:63]:
```suggestion
    for i in range(len(args.in_dwis)):
```

File: scripts/scil_concatenate_dwi.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        all_bvecs.append(bvecs)
+    all_bvals = np.concatenate(all_bvals)
+    all_bvecs = np.concatenate(all_bvecs)
+
+    ref_dwi = nib.load(args.in_dwi[0])
+    all_dwi = np.zeros(ref_dwi.shape[0:3] + (total_size,),
+                       dtype=args.data_type)
+    last_count = ref_dwi.shape[-1]
+    all_dwi[..., 0:last_count] = ref_dwi.get_fdata()
+    for i in range(1, len(args.in_dwi)):
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:78]:
```suggestion
    for i in range(1, len(args.in_dwis)):
```

File: scripts/scil_concatenate_dwi.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        curr_dwi = nib.load(args.in_dwi[i])
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:79]:
```suggestion
        curr_dwi = nib.load(args.in_dwis[i])
```

File: scripts/scil_concatenate_dwi.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        if curr_dwi.shape != ref_dwi.shape:
+            raise ValueError('All DWI must have the same shape.')
+        curr_size = curr_dwi.shape[-1]
+        all_dwi[..., last_count:last_count+curr_size] = curr_dwi.get_fdata()
+
+    np.savetxt(args.out_bvals, all_bvals, '%d')
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:85]:
```suggestion
    np.savetxt(args.out_bval, all_bvals, '%d')
```

File: scripts/scil_concatenate_dwi.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    np.savetxt(args.out_bvecs, all_bvecs.T, '%0.15f')
```
Comment (arnaudbore) [inline scripts/scil_concatenate_dwi.py:86]:
```suggestion
    np.savetxt(args.out_bvec, all_bvecs.T, '%0.15f')
```

File: scripts/scil_concatenate_dwi.py (hunk 2 above)
Comment (frheault) [reply scripts/scil_concatenate_dwi.py]: I didnt do that one, the reason is nibabel force a float. BUT a super common dtype for dwi is int16
So dwi will be loaded in float32 and then converted to either the first one (or args.data_type)

**Review by arnaudbore — APPROVED** [review]

**Pull request merged**